TFNetworkRecLayer.py

```diff
@@ -5293,56 +5293,73 @@ def post_process_final_rec_vars_outputs(self, rec_vars_outputs, seq_len):
 
   Assumes one source input with a time dimension if outside a :class:`RecLayer`.
-  By default ("from" key not provided), it would either use "data", or ":i".
   With `add_to_input`, it will calculate `x + input`.
+  The output shape is the same as the input in all cases.
 
   The positional encoding is the same as in Tensor2Tensor.
   See :func:`TFUtil.get_positional_encoding`.
   """
   layer_class = "positional_encoding"
   recurrent = True
 
-  def __init__(self, add_to_input=False, constant=-1, **kwargs):
+  def __init__(self, add_to_input=False, constant=-1, offset=None, **kwargs):
     """
     :param bool add_to_input: will add the signal to the input
     :param int constant: if positive, always output the corresponding positional encoding.
+    :param None|LayerBase offset: Specify the offset to be added to positions. Expect shape (batch, time) or (batch,).
     """
     super(PositionalEncodingLayer, self).__init__(**kwargs)
     assert len(self.sources) == 1, "%s: expect a single source" % self
     source = self.input_data
+    if self.output.batch_dim_axis:
+      assert self.output.batch_dim_axis == source.batch_dim_axis
+    if self.output.time_dim_axis:
+      assert self.output.time_dim_axis == source.time_dim_axis
+    if offset:  # Assumptions on the shape of `offset`.
+      assert offset.output.batch_dim_axis == 0, "Expect batch_dim_axis to be 0 for offset."
+      if offset.output.time_dim_axis is not None:
+        assert offset.output.time_dim_axis == 1, "Expect time_dim_axis to be 1 for offset."
     if add_to_input:
       assert source.dim == self.output.dim
     from TFUtil import get_positional_encoding
     if source.have_time_axis():
       length = tf.shape(source.placeholder)[source.time_dim_axis]
       if constant > -1:
-        position = constant * tf.ones([length], tf.int32)
-        signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (len,n_out)
+        position = constant * tf.ones([1, length], tf.int32)  # (1, len).
+        if offset:
+          position += offset.output.placeholder  # (batch, len).
+        signal = get_positional_encoding(num_channels=self.output.dim, position=position)  # (1, len) or (batch, len).
       else:
-        signal = get_positional_encoding(num_channels=self.output.dim, length=length)  # (len,n_out)
+        position = tf.range(length)  # (len,).
+        position = tf.expand_dims(position, axis=0)  # (1, len).
+        if offset:
+          position += offset.output.placeholder
+        signal = get_positional_encoding(num_channels=self.output.dim, position=position)
+      # signal has shape (1,len,n_out) or (batch,len,n_out).
+      if source.batch_dim_axis == 1:
+        assert source.time_dim_axis == 0,\
+          "Unexpected input form: batch_dim_axis 1 and time_dim_axis %s." % source.time_dim_axis
+        signal = tf.transpose(signal, [1, 0, 2])
```

Comment:
Better avoid the `tf.transpose`, and construct it with the right shape/axes in the first place.

Reply:
Done.